crates/formality-prove/src/prove/is_local.rs

```diff
@@ -37,24 +37,166 @@ use crate::{
 // `Vec<Foo>` is not. `LocalType<ForeignType>` is local. Type aliases and trait
 // aliases do not affect locality.
 
-/// True if `goal` may be remote. This is
-pub fn may_be_remote(decls: Decls, env: Env, assumptions: Wcs, goal: TraitRef) -> Set<Constraints> {
+judgment_fn! {
+    /// True if `goal` may be implemented in a crate that is not the current one.
+    /// This could be a downstream crate that we cannot see, or it could be a future
+    /// (semver-compatible) version of an upstream crate.
+    ///
+    /// Note that per RFC #2451, upstream crates are not permitted to add blanket impls
+    /// and so new upstream impls for local types cannot be added.
+    pub fn may_be_remote(
+        decls: Decls,
+        env: Env,
+        assumptions: Wcs,
+        goal: TraitRef,
+    ) => () {
+        debug(assumptions, goal, decls, env)
+
+        (
+            (may_be_downstream_trait_ref(decls, env, assumptions, goal) => ())
+            --- ("may be defined downstream")
+            (may_be_remote(decls, env, assumptions, goal) => ())
+        )
+
+        (
+            // In principle this rule could be removed and preserve soundness,
+            // but then we would accept code that is very prone to semver failures.
+            (may_not_be_provable(is_local_trait_ref(decls, env, assumptions, goal)) => ())
+            --- ("may be added by upstream in a minor release")
+            (may_be_remote(decls, env, assumptions, goal) => ())
+        )
+    }
+}
+
+judgment_fn! {
+    /// True if an impl defining this trait-reference could appear in a downstream crate.
+    fn may_be_downstream_trait_ref(
+        decls: Decls,
+        env: Env,
+        assumptions: Wcs,
+        goal: TraitRef,
+    ) => () {
+        debug(goal, assumptions, env, decls)
+
+        (
+            // There may be a downstream parameter at position i...
+            (&goal.parameters => p)
+            (may_be_downstream_parameter(&decls, &env, &assumptions, p) => ())
+            --- ("may_be_downstream_trait_ref")
+            (may_be_downstream_trait_ref(decls, env, assumptions, goal) => ())
+        )
+    }
+}
+
+judgment_fn! {
+    fn may_be_downstream_parameter(
+        decls: Decls,
+        env: Env,
+        assumptions: Wcs,
+        parameter: Parameter,
+    ) => () {
+        debug(parameter, assumptions, env, decls)
+
+        (
+            // existential variables *could* be inferred to downstream types; depends on the substitution
+            // we ultimately have.
+            --- ("type variable")
+            (may_be_downstream_parameter(_decls, _env, _assumptions, TyData::Variable(Variable::ExistentialVar(_))) => ())
+        )
+
+        // If we can normalize `goal` to something else, and that normalized
+        // form may be downstream.
```

Comment:
```
        // If `parameter` is an alias which refers a type which may be
        // from a downstream crate, it may be normalized to that type,
        // so recurse into it.
        //
        // We only do so if the alias cannot be normalized to a type which
        // is definitely not a downstream parameter.

```